### DISCUSSION TITLE
Ternary operators in EE templating language

### DISCUSSION THREAD

**Discussion by robsonsobral** [ExpressionEngine/ExpressionEngine#3106] [Feature Requests]
It's an idea just occurred to me.

To check for an available field and fallback to another can create long code:

```
{if summary}{summary}{if:else}{body:limit characters="280" end_char="…"}{/if}
```

It could be easier to have an Elvis operator, the `?:`, which works like the `||` on JS:

```
{summary ?: body:limit characters="280" end_char="…"}
```

**Comment by bsrweb** [comment]
I'd prefer code that's more obvious to understand what's going on, even it takes a little longer to type. While the suggested syntax may be common knowledge to seasoned developers, there are a lot of non-developers using EE (like designers) specifically because they don't need to know these things. Just my 2 cents...

**Comment by aasun** [comment]
Unless this type of approach would substantially reduce the conditional load on the template (improving performance), I'd rather have a more human-readable code snippet (like the current conditional code).

**Comment by PaulBaileyAtChan** [comment]
Not a crucial thing, but it would be an elegant little addition, and there are places I would use it. Not totally convinced by the idea that it's too much of a techie thing. It's not that much more obscure than something like `!=`, for example, and a more explicit `{if}{if:else}` structure would always be possible. This would be adding to that, not replacing it.

**Comment by robsonsobral** [comment]
In my opinion, long conditionals are less readable. It's more like a matter of experience, I guess.

The idea came to me because I'm coding within a `href` attribute, which forbids the indentation.

**Comment by blue-dreamer** [comment]
Thinking on this one I do prefer the more verbose way EE does conditionals:

- much easier for newbies to pick up (verbose vs "cryptic")
- I'd argue verbose is easier to debug for the majority (not everyone who use EE is a programmer)
- if it ain't broke don't fix it
- not sure about how your Elvis (has left the building) method would work with multiple/nested conditionals?

Would this be an enhancement or a full replacement?

**Comment by PaulBaileyAtChan** [comment]
Good grief, people. This isn't being suggested as a replacement for basic `{if}{if:else}`. How could it possibly be that? It would be a nifty little alternative for a specific subset of conditions, for the people who like it and want to use it. (Not everyone who uses EE is not a programmer.)

**Comment by bsrweb** [comment]
Would this be something that could be easily made into a plugin? Seems like this might be the best option, something we can use if we want it. I agree with @aasun: not sure if adding more code to the template parser to replicate existing functionality is a prudent policy unless it significantly improves EE's performance or if it's something that 90% of developers want to improve their performance for building templates (which could be the case here).

**Comment by litzinger** [comment]
If the current template parser is going to to be around for a few more years, yes absolutely add something like this. If the plan is to replace the parser with Twig within the next year or two, then maybe it isn't with the time to add it.

The argument that it's not as clear is a weak one because every other language that supports it also supports the longer, more terse conditionals as well. Pick your poison. If a new developer sees something like this and doesn't understand it then it's a learning opportunity. 

Not adding features like this stunts EE's growth. Every time I show EE template code to a PHP dev their eyes roll when they see `{if:else}` and `{if:elseif}`. EE can cater to both audiences equally.

Edit: we already have the `*=`, `~`, and other regex based conditional logic, which is arguably more confusing than this proposal because it is unique to EE. This is a suggestion for something that is already a very common and widely accepted syntax.

**Comment by robsonsobral** [comment]
I praise the consideration for non devs the community has. However, I've seen this often translated into resistance to features a more experienced developer could use.

Maybe because English isn't my first language, since I started to code, I think ternaries are much easier to understand than to redefine variables within conditionals, just to pick an example I usually see on discussions.

**Comment by PaulBaileyAtChan** [comment]
(Total OT aside here, but if a goal for EE is to reach out to non-programmers, avoiding more expressive template syntax doesn't help one iota. Providing some pre-built templates for common website uses *might*.)

**Comment by litzinger** [comment]
Another point to make... I've seen some _really_ shit show templates in my day that are 100x more confusing than any form of conditional you can think of, and this is coming from a seasoned EE and PHP developer. A `?:` operator in some of the templates I've seen is the last thing a new developer would be confused about.

**Comment by bsrweb** [comment]
Good and fair points, @litzinger. But, please don't be too easily dismissive of what or what might not be confusing in EE's syntax to a non-developer (like me). I and a lot of other EE devs here come from that background and this is the feedback we're giving EEHarbor. And we've all seen shitty, confusing templates—heck, I've built some of them—but that's a bit of a straw man argument in this case.

**Comment by ssssct** [comment]
> the ?:, which works like the || on JS:

If you felt you had to explain what it does, you might admit it's not so obvious or intuitive.
I feel the `||` or even `OR` would be easier on the eyes of most people, but it surely is a matter of point of view.

**Comment by robsonsobral** [comment]
Oh, I totally agree! As a JS dev, I prefer `||`! I suggested `?:` because I thought it could be easier to parse.

```
{summary || body}
```

**Comment by wagonheart** [comment]
i like the idea, i've definitely needed single line output and end up with blocks of code that look like:
`{if summary}{!--`
`--}{summary}{!--`
`--}{if:else}{!--`
`--}{body:limit characters="280" end_char="…"}{!--`
`--}{/if}`
this is obviously a simplistic version, but same idea.

using expressionengine since early v2 has definitely made programming feel more accessible to me, as mostly a front end dev, and feel the introduction through ee to these concepts might open up more possibilities/ideas to non programmers using ee.

`||` and `OR` would be more in keeping with ee's existing syntax, but nice idea

but as mentioned, not as a complete replacement

**Comment by litzinger** [comment]
After thinking about this more I came to the same conclusion as @robsonsobral and @ssssct that the `||` operator might make more sense in this case. The `?:` operator would only make sense if the parser also supported a full ternary, since the Elvis operator is short hand for it. So unless it supports `{condition ? valueA : valueB}` I think I'd prefer to use `||` instead.

Another thing to consider in either case, is if it supports a full ternary, would it also support a ternary in a ternary? :) This is certainly where readability comes into play and I think universally people agree a nested ternary is just bad.

If the `||` operator is used, does the support stop at `{valueA || valueB}` or does it support `{valueA || valueB || valueC}`?

About my earlier comments, I didn't mean to be dismissive. I too try to take a proactive approach and making sure people new to a project can understand what is going on. Usually that is done by adding contextual comments explaining why something that might appear abnormal is present in the code. I personally think something like a shorthand variable ternary is so low on the list of complex things to understand for a non-programmer or a programmer new to the EE. For example the `dynamic="no"` parameter has caused so much confusion. Or understanding EE's routing (template based URLs vs Pages/Structure). Or understanding the parse order basics. The `{preload_replace:var}` variable syntax is entirely unique to EE. If someone sees it and doesn't understand it, they look for it in the docs. If someone encounters a `{valueA || valueB}` syntax and doesn't understand it, they should be able to learn about it in the docs.

As someone mentioned earlier this is a suggestion to add something new that can make template code cleaner. The old `{if}` logic would still be an option for those who choose to use it.

**Comment by robsonsobral** [comment]
@litzinger 's comments about Elvis origin make a perfect argument. It only makes sense in case of full ternaries available.

Bye, Elvis!

**Comment by PaulBaileyAtChan** [comment]
(Elvis has left the building?)

For what it's worth, I'm not a fan of using `||` here, for a couple of reasons.

The extended meaning of `||` in JS as returning the first operand that evaluates to true is sort of a hacky confusion with a pure logical OR, and that lack of clarity doesn't help. (So, yes, I'm saying it would be conceptually messy here because I think it's conceptually messy in JS.)

Also, it relies on a left-to-right evaluation order, which is an implementation detail, and *not* something that's inherent in the operator. For example, with a pure logical OR, `(x || y)` and `(y || x)` should behave identically. That's not what we're trying to do here. We're talking about something with a very clear left-to-right order. The following are very different:

```
if x exists, then use x, otherwise use y
if y exists, then use y, otherwise use x

```
A traditional ternary operator has a clear left-to-right order built in, not just as an implementation detail, so it's clearer here. Whether it's Elvis or not, the ternary operator is IMO the right *model* for this suggestion.

**Comment by litzinger** [comment]
Thanks for clarifying that. I'm not a JS ninja so that actually wasn't clear to me.

**Comment by blue-dreamer** [comment]
Aside - @PaulBaileyAtChan the "Elvis has left the building" sound bite has been used in various places over the years, one I remember was a 1990's Windows DTP program that when you looked at the "about" dialogue the sound bite would play - kind of like a software Easter egg. A bit like Winamps "It really whips the lamas ass" https://www.youtube.com/watch?v=HaF-nRS_CWM

**Comment by ic360-com** [comment]
I like the idea of the above suggested code (or a simplified EE-style form of it, if that's even possible) - in addition to retaining the current {if}{if:else}{/if} tags.
I've also often thought something like this would be useful:
`{summary wrap="<strong></strong>" after="<br><br>"}`
So if "summary" exists, wrap it with &lt;strong&gt; tags and output a couple of line breaks after it.

**Comment by liljefelt** [comment]
(Sorry for the excavation)

I had a similar idea last night, but only for a single default value, ie a shorthand for this;

`{if variable} {variable} {if:else}  {default_value} {/if}`

My imaginary syntax would look something like this:

`{variable ? "default_value"}`

`{segment_3 ? "default_value"}`

`{embed:url_title ? "default_value"}`

`{layout:url_title ? "default_value"}`


Has there been any progress on this?
